### PR TITLE
Remove uses of plugins in the snapshot

### PR DESCRIPTION
**Pull request by swgillespie**
Fixes https://github.com/pulumi/pulumi/issues/2576 by basically deprecating the plugin manifest.

For historical context, the plugin manifest was for a long time the source of truth about what plugins were necessary to operate upon a Pulumi program. For updates and previews, Pulumi would load plugins on-demand as `RegisterResource` calls came from the language host. As Pulumi loaded those plugins lazily, it would record in the snapshot manifest that it loaded the plugin and the filepath of the plugin on-disk. When running a delete, Pulumi would consult the manifest to obtain a list of plugins that were used the last time we did an update, and therefore the set of plugins that we would need in order to *delete* the stack.

Today, all providers are reified in the snapshot explicitly as "first-class providers". These providers are resources in every sense. All resources in Pulumi record the provider that created it, and the provider itself records its version. When doing a delete, we don't actually need the snapshot manifest anymore; we can gather the complete set of provider plugins required to do a deletion because all providers are reified explicitly in the snapshot.

However, the manifest was still in use, causing issues such as https://github.com/pulumi/pulumi/issues/2576. Pulumi still recorded plugin loads, but it 1) never clears the list of plugins and 2) loads all plugins in the manifest, even if they aren't needed by the program. The combination of 1) and 2) means that Pulumi stacks forever accumulated a bunch of random plugins in the manifest. Furthermore, the plugins located in the snapshot manifest recorded a filepath that the engine would use to try and load the plugin again, which is almost guaranteed to be garbage when running updates across multiple machines.

This commit deprecates the plugin manifest. It does so in three ways:

1. The `plugin.Events` callback interface is removed entirely from the engine. The fact that it ever existed was a kludge to work around the fact that we wanted to separate persistence logic from the core engine but the engine needed to persist plugin loads. Without the manifest, it is no longer necessary.
2. When computing the set of required plugins, the engine ignores the snapshot manifest entirely.
3. When persisting snapshots, the snapshot manager doesn't write out the plugin manifest.

### CODE REVIEW

File: pkg/backend/snapshot.go [hunk 1]
```diff
@@ -507,7 +496,7 @@ func (sm *SnapshotManager) snap() *deploy.Snapshot {
 	manifest := deploy.Manifest{
 		Time:    time.Now(),
 		Version: version.Version,
-		Plugins: sm.plugins,
+		// Plugins: sm.plugins, - Explicitly dropped, since we don't use the plugin list in the manifest anymore.
```
Comment (ellismg) [inline pkg/backend/snapshot.go]: Is there a compelling reason for retaining the `Plugins` field of `deploy.Manifest` at all?

Comment (swgillespie) [reply pkg/backend/snapshot.go]: I don't think so - as long as this is an engine-internal type (which I'm pretty sure it is?) it should be fine.

Comment (swgillespie) [reply pkg/backend/snapshot.go]: Turns out there is a compelling reason:

When the engine encounters a stack whose snapshot doesn't have any default provider references, it injects them as necessary:

https://github.com/pulumi/pulumi/blob/0550f71a35df8bce291d299f5b4c653968f1fe8c/pkg/resource/deploy/plan.go#L111-L115

In order to do this, it uses the manifest to figure out what version of the plugin to inject into the first-class provider resource that it synthesizes. I don't think we can safely remove this code without breaking users using this CLI to update a very old stack.

Comment (ellismg) [reply pkg/backend/snapshot.go]: Makes sense!

**Pull request merged**